### PR TITLE
closes #8895 add std/once

### LINKED ISSUE #8895: race condition for system.once?
`system.once` doesn't seem safe in multithreaded context:
```
template once*(body: untyped): untyped =
  var alreadyExecuted {.global.} = false
  if not alreadyExecuted:
    alreadyExecuted = true
    body
```

## see also
* https://stackoverflow.com/questions/8412630/how-to-execute-a-piece-of-code-only-once
* http://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

### PR DESCRIPTION
**Pull request by ringabout**
fix #8895
add thread-safe once to std/once, for ORC.

### CODE REVIEW

File: lib/std/once.nim [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+type
+  Once* = object
+    finished: Atomic[bool]
+    lock: Lock
+
+proc initOnce*(once: var Once) =
+  once.finished.store(false)
+  initLock(once.lock)
+
+template once*(alreadyExecuted: Once, body: untyped) =
```
Comment (timotheecour) [inline lib/std/once.nim]: can a 0 arg version work eg:

```nim
var count = 0
for i in 0..<3:
  once: inc count
doAssert count == 1
```

File: lib/std/once.nim [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import atomics
+import locks
```
Comment (timotheecour) [inline lib/std/once.nim]: what about re-entrant code? do we need a `ronce / reentrantOnce` (rlocks), or can `once` be made re-entrant safe without overhead?

File: lib/std/once.nim [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import atomics
```
Comment (timotheecour) [inline lib/std/once.nim]: this doesn't work:
```nim
when true:
  import std/once
  var block1: Once
  initOnce(block1)
  var thr: array[0..4, Thread[void]]
  proc threadFunc() {.thread.} =
    var count = 0
    for i in 1 .. 10:
      once(block1):
        echo "in once"
  for i in 0..high(thr):
    createThread(thr[i], threadFunc)
  joinThreads(thr)
```

depending on the runs, you'll get anywhere from 1 to 4 `in once`

Comment (ringabout) [reply lib/std/once.nim]: fixed it

Comment (ringabout) [reply lib/std/once.nim]:
```nim
block:
  var thr: array[0..4, Thread[void]]
  proc threadFunc() {.thread.} =
    var block1: Once
    initOnce(block1)
    for i in 1 .. 10:
      once(block1):
        echo "in once"
  for i in 0..high(thr):
    createThread(thr[i], threadFunc)
  joinThreads(thr)
```
This also works for me: print `in once` five times.

File: lib/std/once.nim (hunk 1 above)
Comment (ringabout) [reply lib/std/once.nim]: No, it doesn't work because of `lock`. You need to maintain a var lock for each instance. It hard to use `{.global.}` to achieve this(maybe pure Nim version lock could improve this).

**Review by dom96 — COMMENTED** [review]
Instead of implementing a new module for this, can't we fix `system.once` instead?

File: tests/stdlib/tonce.nim [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  block:
+    var thr: array[0..4, Thread[void]]
+    var block1: Once
+    var count = 0
+    initOnce(block1)
+    proc threadFunc() {.thread.} =
+      for i in 1 .. 10:
+        once(block1):
+          inc count
+    for i in 0..high(thr):
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim]: more idiomatic:
```nim
for t in mitems(thr):
  createThread(t, threadFunc)
```
(and Example code in https://github.com/nim-lang/Nim/blob/73299b048f57838d83a3a8159909386a0905e553/lib/system/threads.nim#L41 should be updated as well in future work)

ditto below

**Comment by timotheecour** [comment]
> Instead of implementing a new module for this, can't we fix system.once instead?

if it can be done transparently for user code and it doesn't add new dependencies to system.nim, yes, but with the way it's done in this PR at least, this isn't the case as it requires an extra parameter `Once` so would require users to change their code anyways. So it's better to delegate this to another module and avoid bloating system.nim further. 

- [ ] However existing `system.once` should be marked as deprecated in the same PR, just like it was the case for `addQuitProc` in https://github.com/nim-lang/Nim/pull/14342

File: tests/stdlib/tonce.nim [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    proc threadFunc() {.thread.} =
+      var block1: Once
+      initOnce(block1)
+      for i in 1 .. 10:
+        once(block1):
+          inc count
+    for i in 0..high(thr):
+      createThread(thr[i], threadFunc)
+    joinThreads(thr)
+    doAssert count == 5
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim]: `else: static doAssert false`
(prevents silent misuse)

File: tests/stdlib/tonce.nim [hunk 6]
```diff
@@ -0,0 +1,34 @@
+discard """
+  cmd: "nim c -r --threads:on $options $file"
+  matrix: "-d:caseA; -d:caseB"
+"""
+import std/once
+
+when defined(caseA):
+  block:
+    var thr: array[0..4, Thread[void]]
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim]: more idiomatic:
* `var thr: array[5, Thread[void]]`
* and as https://github.com/nim-lang/Nim/pull/16192/files#r540547170 upstream code in threads.nim should be updated

File: tests/stdlib/tonce.nim (hunk 5 above)
Comment (timotheecour) [inline tests/stdlib/tonce.nim]: `doAssert count == thr.len`

File: tests/stdlib/tonce.nim [hunk 7]
```diff
@@ -0,0 +1,34 @@
+discard """
+  cmd: "nim c -r --threads:on $options $file"
+  matrix: "-d:caseA; -d:caseB"
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim]: here you actually don't need matrix for that since you can combine both:

```nim
block:
  var thr: array[0..4, Thread[void]]
  ...

block:
  var thr: array[0..4, Thread[void]]
  ...
```

but instead, use matrix for:
```
matrix: "--threads:off; --threads:on"
```

and adjust code to work with and without threads with `compileOption`. There are ways to do it without any code duplication.

File: tests/stdlib/tonce.nim [hunk 8]
```diff
@@ -0,0 +1,34 @@
+discard """
+  cmd: "nim c -r --threads:on $options $file"
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim:2]: add `# xxx: support vm+js`
(right now doesn't work, not even considering webworkers)

File: tests/stdlib/tonce.nim [hunk 9]
```diff
@@ -0,0 +1,34 @@
+discard """
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim:1]: js needs to be supported because `system.once` already works in js (not considering edge case of webworkers)

(and leave a TODO for supporting VM because VM already doesn't work with system.once)

```nim
when true:
  var count = 0
  var countCT {.compileTime.} = 0
  proc fn(n: int) =
    once:
      when nimvm:
        countCT.inc
      else:
        count.inc
      echo (n,)
    if n > 1:
      fn(n-1)
  proc main() =
    fn(5)
    fn(4)
    when nimvm:
      echo countCT
      # doAssert countCT == 1 # fails
    else:
      doAssert count == 1
  static: main()
  main()
```

File: tests/stdlib/tonce.nim [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+elif defined(caseB):
+  block:
+    var thr: array[0..4, Thread[void]]
+    var count = 0
+    proc threadFunc() {.thread.} =
+      var block1: Once
+      initOnce(block1)
+      for i in 1 .. 10:
+        once(block1):
+          inc count
```
Comment (timotheecour) [inline tests/stdlib/tonce.nim]: we also need a test where the code in once block raises

File: lib/std/once.nim [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    doAssert count == 1
+
+  if not cond.finished.load(moAcquire):
+    withLock cond.lock:
+      # TODO load a value without atomic
+      if not cond.finished.load(moRelaxed):
+        try:
+          body
+        finally:
+          cond.finished.store(true, moRelease)
```
Comment (timotheecour) [inline lib/std/once.nim]: I'm wondering whether we should customize whether `cond.finished.store(true, moRelease)` happens in a `finally` block or only after a successful `body` that didn't raise.

there are use cases for both.
eg:
```nim
var db: MyDataBase
proc fn=
  once(Once, retryOnFailure = true):
    db = newMyDataBaseFlaky() # suppose this fails 10% of the times
  # we want to make sure `db` is always initialized here
  db.call() # if `newMyDataBase` raised in another thread, `db` will be nil
```

without `retryOnFailure` flag, how would you emulate `retryOnFailure` in user code calling `std/once` ?

**Comment by juancarlospaco** [comment]
Should be named `std/onces`, maybe it can have more than one `once` overload, etc.

File: lib/std/once.nim [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import std/[atomics, locks]
+
+
+type
+  Once* = object    ## A object that ensures a block of code is executed once.
+    finished: Atomic[bool]
+    lock: Lock
+
+proc initOnce*(once: var Once) =
```
Comment (juancarlospaco) [inline lib/std/once.nim]: Can be `func` ?.

**Comment by konsumlamm** [comment]
> Should be named `std/onces`, maybe it can have more than one `once` overload, etc.

`onces` doesn't make much sense, it's not even a word (once is an adverb, not a subject).

**Comment by juancarlospaco** [comment]
What word is `jsffi`, `cgi`, `coro`, etc :shrug:

**Comment by planetis-m** [comment]
My gripe with this implementation is that a user needs to init once explicitly. This is unusual and can cause UB if its forgotten. An easy way is using a [spinlock](https://codereview.stackexchange.com/questions/117468/stdonce-flag-and-stdcall-once-implementation) More advanced implementations are in [rust](https://github.com/rust-lang/rust/tree/master/library/std/src/sync) and [abseil](https://github.com/abseil/abseil-cpp/blob/master/absl/base/call_once.h)
